include/nbl/builtin/glsl/virtual_geometry/virtual_attribute_fetch.glsl

```diff
@@ -0,0 +1,180 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_BUILTIN_GLSL_VIRTUAL_GEOMETRY_VIRTUAL_ATTRIBUTE_FETCH_INCLUDED_
+#define _NBL_BUILTIN_GLSL_VIRTUAL_GEOMETRY_VIRTUAL_ATTRIBUTE_FETCH_INCLUDED_
+
+#include <nbl/builtin/glsl/format/decode.glsl>
+#include <nbl/builtin/glsl/virtual_geometry/virtual_attribute.glsl>
+
+//TODO: R8, R8G8, R8G8B8, RGBA8, R16, RGB16
+// float nbl_glsl_VG_attribFetch_R8_UNORM(in nbl_glsl_VG_VirtualAttributePacked_t attr, in uint vertexID)
+// {
+//   return nbl_glsl_VG_attribFetch1f(attr, vertexID);
+// }
+// vec4 nbl_glsl_VG_attribFetch_RGBA8_SFLOAT(in nbl_glsl_VG_VirtualAttributePacked_t attr, in uint vertexID)
+// {
+// #ifdef _NBL_VG_USE_SSBO
+//   return unpackUnorm4x8(nbl_glsl_VG_attribFetch1u(attr, vertexID))
+// #else
+// #endif
+// }
+
+#if (defined(_NBL_VG_USE_SSBO) && defined(_NBL_VG_USE_SSBO_UVEC4)) || _NBL_VG_FLOAT_BUFFERS_COUNT
+vec4 nbl_glsl_VG_attribFetch_RGBA32_SFLOAT(in nbl_glsl_VG_VirtualAttributePacked_t attr, in uint vertexID)
+{
+  return nbl_glsl_VG_attribFetch4f(attr, vertexID);
+}
+#endif
+
+#if (defined(_NBL_VG_USE_SSBO) && defined(_NBL_VG_USE_SSBO_UVEC3)) || _NBL_VG_FLOAT_BUFFERS_COUNT
+vec3 nbl_glsl_VG_attribFetch_RGB32_SFLOAT(in nbl_glsl_VG_VirtualAttributePacked_t attr, in uint vertexID)
+{
+  return nbl_glsl_VG_attribFetch3f(attr, vertexID);
+}
+#endif
+
+#if (defined(_NBL_VG_USE_SSBO) && defined(_NBL_VG_USE_SSBO_UVEC2)) || _NBL_VG_FLOAT_BUFFERS_COUNT
+vec2 nbl_glsl_VG_attribFetch_RG32_SFLOAT(in nbl_glsl_VG_VirtualAttributePacked_t attr, in uint vertexID)
+{
+  return nbl_glsl_VG_attribFetch2f(attr, vertexID);
+}
+#endif
+
+#if (defined(_NBL_VG_USE_SSBO) && defined(_NBL_VG_USE_SSBO_UINT)) || _NBL_VG_FLOAT_BUFFERS_COUNT
+float nbl_glsl_VG_attribFetch_R32_SFLOAT(in nbl_glsl_VG_VirtualAttributePacked_t attr, in uint vertexID)
+{
+  return nbl_glsl_VG_attribFetch1f(attr, vertexID);
+}
+#endif
+
+#if (defined(_NBL_VG_USE_SSBO) && defined(_NBL_VG_USE_SSBO_UVEC2)) || _NBL_VG_FLOAT_BUFFERS_COUNT
+vec4 nbl_glsl_VG_attribFetch_RGBA16_SFLOAT(in nbl_glsl_VG_VirtualAttributePacked_t attr, in uint vertexID)
+{
+#ifdef _NBL_VG_USE_SSBO
+  uvec2 packed = nbl_glsl_VG_attribFetch2u(attr, vertexID);
+  vec2 xy = unpackHalf2x16(packed.x).yx;
+  vec2 zw = unpackHalf2x16(packed.y).yx;
```

Comment:
why on earth are you swapping the order of components!?

Reply:
http://docs.gl/sl4/unpackHalf2x16

> The first component of the vector is obtained from the 16 least-significant bits of v; the second component is obtained from the 16 most-significant bits of v.

Reply:
Yes, so why the element swap?

Reply:
i see, my bad